docs/topics/addons.rst

```diff
@@ -0,0 +1,348 @@
+developers.
+
+
+Activating and configuring add-ons
+==================================
+
+Add-ons and their configuration live in Scrapy's
+:class:`~scrapy.addons.AddonManager`. During Scrapy's start-up process, and
+only then, the add-on manager will read a list of enabled add-ons and their
+configurations from your ``ADDONS`` setting.
+
+The ``ADDONS`` setting is a dict in which every key is an addon class or its
+import path and the vaoue is its priority.
+
+The configuration of an add-on, if necessary at all, is stored as a dictionary
+setting whose name is the uppercase add-on name.
+
+This is an example where two add-ons (in this case with one requiring no
+configuration) are enabled/configured in a project's ``settings.py``::
+
+    ADDONS = {
+        'path.to.someaddon': 0,
+        path.to.someaddon2: 1,
+    }
+
+    SOMEADDON = {
+        'some_config': True,
+    }
+
+Enabling and configuring add-ons within Python code
+---------------------------------------------------
+
+The :class:`~scrapy.addons.AddonManager` will only read from Scrapy's settings
+*at the beginning* of Scrapy's start-up process.
+Afterwards, i.e. as soon as the :class:`~scrapy.addons.AddonManager` is
+populated, changing the ``ADDONS`` setting or any of the add-on
+configuration dictionary settings will have no effect.
+
+If you want to enable, disable, or configure add-ons in Python code, for example
+when writing your own add-on, you will have to use the
+:class:`~scrapy.addons.AddonManager`. You can access the add-on manager through
+either ``crawler.addons`` or, if you are writing an add-on, through the
+``addons`` argument of the :meth:`update_addons` callback. The add-on manager
+provides many useful methods and attributes to facilitate interacting with the
+add-ons framework, e.g.:
+
+* an :meth:`~scrapy.addons.AddonManager.add` method to load add-ons,
+* the :attr:`~scrapy.addons.AddonManager.enabled` list of enabled add-ons,
+* :meth:`~scrapy.addons.AddonManager.enable` and
+  :meth:`~scrapy.addons.AddonManager.disable` methods,
+* the :attr:`~scrapy.addons.AddonManager.configs` dictionary which holds the
+  configuration of all add-ons.
+
+In this example, we ensure that the ``httpcache`` add-on is loaded, and that
+its ``expiration_secs`` configuration is set to ``60``::
+
+    # addons is an instance of AddonManager
+    if 'httpcache' not in addons:
+        addons.add('httpcache', {'expiration_secs': 60})
+    else:
+        addons.configs['httpcache']['expiration_secs'] = 60
+
+
+Writing your own add-ons
+========================
+
+Add-ons are (any) Python *objects* that provide Scrapy's *add-on interface*.
+The interface is enforced through ``zope.interface``. This leaves the choice of
+Python object up the developer. Examples:
+
+* for a small pipeline, the add-on interface could be implemented in the same
+  class that also implements the ``open/close_spider`` and ``process_item``
+  callbacks,
```

Comment:
In my experience this is an anti-pattern to have a single class act like different types of Scrapy components. The issue is that Scrapy would create multiple objects of this class; in this case, one instance would be used as an addon, and another as an extension. This can lead to confusion. If there is any work done during initialization ( even logging), it'd be done twice.